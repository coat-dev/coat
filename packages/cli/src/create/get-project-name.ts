import { prompt } from "inquirer";
import validateNpmPackageName from "validate-npm-package-name";

function sanitizeProjectName(input: string): string {
  return input.trim();
}

function createProjectNameErrorMessage(
  validation: validateNpmPackageName.Result
): string {
  return `The project name must be a valid npm package name: ${[
    ...(validation.warnings ?? []),
    ...(validation.errors ?? []),
  ].join(", ")}`;
}

/**
 * Returns true when a project name is valid for coat.
 * If it is invalid, then a string is returned which can be used as
 * the error message to the user. This format is chosen to be
 * compatible with the inquirer validation methods.
 */
function validateProjectName(projectName: string): string | true {
  const npmValidation = validateNpmPackageName(projectName);
  if (npmValidation.validForNewPackages) {
    return true;
  }
  return createProjectNameErrorMessage(npmValidation);
}

export async function getProjectName(
  projectNameInput: string | undefined
): Promise<string> {
  let projectName = "";
  // The project name must be a valid npm package name and is placed as the name of
  // the resulting package.json file.
  //
  // See official docs: https://docs.npmjs.com/creating-a-package-json-file
  // Also see validate-npm-package-name: https://github.com/npm/validate-npm-package-name
  if (projectNameInput) {
    const sanitizedProjectNameInput = sanitizeProjectName(projectNameInput);
    const validationResult = validateProjectName(sanitizedProjectNameInput);
    if (validationResult === true) {
      projectName = sanitizedProjectNameInput;
    } else {
      console.warn(validationResult);
    }
  }

  // If no project name is specified with the command - or the specified project name is
  // not valid - the user should be prompted for the project name.
  if (!projectName) {
    const answers = await prompt([
      {
        name: "projectName",
        message: "Enter the name of your new project",
        filter: sanitizeProjectName,
        default: projectNameInput,
        validate: validateProjectName,
      },
    ]);
    projectName = answers.projectName;
  }
  return projectName;
}
