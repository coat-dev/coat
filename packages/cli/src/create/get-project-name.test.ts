import { getProjectName } from "./get-project-name";
import { prompt } from "inquirer";

jest.mock("inquirer");

((prompt as unknown) as jest.Mock).mockImplementation(() => ({
  projectName: "prompted-name",
}));

jest.spyOn(console, "warn").mockImplementation(() => {
  // Ignore console messages
});

describe("create/get-project-name", () => {
  test("should take specified project name if it is valid", async () => {
    const projectName = "project-name";
    const result = await getProjectName(projectName);
    expect(result).toBe(projectName);
  });

  test("should automatically trim specified project name without prompting the user", async () => {
    const projectName = "  project-name  ";
    const result = await getProjectName(projectName);
    expect(result).toBe("project-name");
  });

  test("should prompt the user if no project name is specified", async () => {
    const result = await getProjectName(undefined);
    expect(result).toBe("prompted-name");
  });

  test.each`
    projectName
    ${"My-Project"}
    ${"my$project"}
  `(
    "should warn user and prompt if specified project name ($projectName) is invalid",
    async ({ projectName }) => {
      (console.warn as jest.Mock).mockClear();

      const result = await getProjectName(projectName);
      expect(result).toBe("prompted-name");
      expect(console.warn).toHaveBeenCalledTimes(1);
      const warnArguments = (console.warn as jest.Mock).mock.calls[0];
      expect(warnArguments).toMatchSnapshot();
    }
  );
});
