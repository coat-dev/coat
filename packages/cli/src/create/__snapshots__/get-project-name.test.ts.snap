// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create/get-project-name should warn user and prompt if specified project name (My-Project) is invalid 1`] = `
Array [
  "The project name must be a valid npm package name: name can no longer contain capital letters",
]
`;

exports[`create/get-project-name should warn user and prompt if specified project name (my$project) is invalid 1`] = `
Array [
  "The project name must be a valid npm package name: name can only contain URL-friendly characters",
]
`;
